proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,565 @@
+
+::
+
+  -- In UserId module
+  fromNumeric :: Natural -> UserId
+  fromNumeric = UserId
+
+  UserId.123 -- works
+
+  UserId.(-1) -- error: (-1) has type Integer, expected Natural
+
+See https://github.com/ghc-proposals/ghc-proposals/issues/438 for more details.
+
+ByteString
+~~~~~~~~~~
+
+It's a `known issue <https://github.com/haskell/bytestring/issues/140>`_ that ByteString has surprising ``IsString`` behavior, due to ambiguity in how to handle Unicode characters.
+
+With ``QualifiedStrings``, ``bytestring`` could define the following modules:
+
+::
+
+  module Data.ByteString.Qualified.Ascii where
+
+  -- truncates unicode
+  fromString :: String -> ByteString
+  fromString = Char8.pack
+
+  module Data.ByteString.Qualified.Utf8 where
+
+  -- encodes unicode
+  fromString :: String -> ByteString
+  fromString = BS.toStrict . BS.toLazyByteString . BS.stringUtf8
+
+Users would then be forced to decide what behavior they want (and can switch between the two!):
+
+::
+
+  import Data.ByteString qualified as BS
+  import Data.ByteString.Qualified.Ascii qualified as Ascii
+  import Data.ByteString.Qualified.Utf8 qualified as Utf8
+
+  main = do
+    -- [98,108,97,158]
+    print $ BS.unpack Ascii."bla語"
+
+    -- [98,108,97,232,170,158]
+    print $ BS.unpack Utf8."bla語"
+
+Heterogeneous Lists
+~~~~~~~~~~~~~~~~~~~
+
+With ``QualifiedLists``, converting list literals are no longer confined to the list type, enabling list literal syntax for heterogenous lists (aka ``HList``):
+
+::
+
+  module Data.HList.Qualified where
+
+  buildList ::
+    ( (forall a as. f a -> HList f as -> HList f (a ': as))
+      -> HList f '[]
+      -> HList f xs
+    ) -> HList f xs
+  buildList f = f HCons HNil
+
+  pattern FromListCons :: () => xs ~ (x0 ': xs0) => f x0 -> HList f xs0 -> HList f xs
+  pattern FromListCons a b = HCons a b
+
+  pattern FromListNil :: () => xs ~ '[] => HList f xs
+  pattern FromListNil = HNil
+
+  {-# COMPLETE FromListCons, FromListNil #-}
+
+Users could then do
+
+::
+
+  import Data.HList.Qualified qualified as HList
+
+  HList.[Just True, Just 1, Nothing] :: HList Maybe '[Bool, Int, String]
+
+  -- hlist matches all the patterns below
+  case hlist of
+    HList.[Just True, _, Nothing] -> _
+    HList.[_, Just 1, Nothing] -> _
+    Just _ HList.: _ -> _
+
+ByteArray
+~~~~~~~~~
+
+Example of a ``ByteArray`` implementation, which requires knowing the length of the list in advance.
+
+::
+
+  type Builder s = (Int -> MutableByteArray s -> ST s (), Int)
+
+  buildList ::
+    forall a. Prim a =>
+    ( forall s.
+      (a -> Builder s -> Builder s)
+      -> Builder s
+      -> Builder s
+    ) -> ByteArray
+  buildList f = let (go, n) = f cons nil in createByteArray (n * sizeOfType @a) (go 0)
+    where
+      nil :: Builder s
+      nil = (\_ _ -> pure (), 0)
+
+      cons :: Prim a => a -> Builder s -> Builder s
+      cons x (go, n) = (\i arr -> writeByteArray arr i x >> go (i + 1) arr, n + 1)
```

Comment:
Could you give an example of why a user would be writing this? At face value I'm confused why a user would want to use list syntax for constructing a ByteArray, and especially doing so using this `Builder` type?

Reply:
Why wouldn't a user want to use list syntax for ByteArray? ByteArray currently has an `IsList` instance, so presumably _someone_ is using `OverloadedLists` with it. If I was writing code that needed ByteArray, it would be great to write tests like `ByteArray.[1, 2, 3]`

The Builder type is only necessary to avoid the double traversal. If performance isn't an issue, you could write this as
```hs
buildList f =
  createByteArray (length list * sizeOfType @a) $ \arr ->
    mapM_ (uncurry $ writeByteArray arr) (zip [0 ..] list)
  where
    list = f (:) []
```

Reply:
Ah I misunderstood. So it looks like this turns a het list of `Prim a => a` values into a `ByteArray`. It would be good to say that and give an example in the proposal.

Does it actually avoid traversing things twice? Don't you traverse once to get the length and then traverse again to write things to the `ByteArray`? You avoid allocating an intermediate list, but you do allocate things to represent your `go` function.